### ISSUE TITLE
運行bash ./scripts/metaqa/run.sh 的hist1, hist3, hist10皆為0

### ISSUE THREAD

**Issue by JoeySmith1103** [zjunlp/PromptKG#47]
您好，我目前的环境是torch1.10.1 搭配cuda 11.1，目前只有更動run.sh，改为:
CUDA_VISIBLE_DEVICES=0 python main.py  --max_epochs=10  --num_workers=0 \
   --model_name_or_path  t5-base \
   --num_sanity_val_steps 0 \
   --model_class T5KBQAModel \
   --lit_model_class KGT5LitModel \
   --label_smoothing 0.1 \
   --data_class MetaQADataModule \
   --precision 16 \
   --batch_size 16 \
   --check_val_every_n_epoch 2 \
   --dataset metaQA \
   --k_hop 1 \
   --eval_batch_size 64 \
   --max_seq_length 64 \
   --max_entity_length 128 \
   --lr 5e-5

也就是更改 CUDA_VISIBLE_DEVICES、num_workers、batch_size以及eval_batch_size，但输出的结果都为0，请问是什么步骤出错了吗？

**Comment by zxlzr** [comment]
请问您用没改过的脚本跑的结果正常吗？

**Comment by JoeySmith1103** [comment]
可能是GPU只有一个的问题，num_workers的部分我除了0以外的值在训练都会卡住没有进度。

**Comment by CheaSim** [comment]
可能需要您使用`google-t5/t5-base` 加载模型，进行实验。

**Comment by JoeySmith1103** [comment]
已更改为google-t5/t5-base模型，但输出结果仍为0，以下为输出结果:
Global seed set to 7
Downloading: 100%|███████████████████████████████████████| 1.21k/1.21k [00:00<00:00, 3.66MB/s]Downloading: 100%|██████████████████████████████████████████| 792k/792k [00:00<00:00, 963kB/s]Downloading: 100%|███████████████████████████████████████| 1.39M/1.39M [00:00<00:00, 1.40MB/s]Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
<data.data_module.MetaQADataModule object at 0x7f71693b2fd0>
Downloading: 100%|█████████████████████████████████████████| 892M/892M [01:53<00:00, 7.89MB/s]tokenize the prefix tree: 100%|██████████████████████| 43234/43234 [00:00<00:00, 48702.13it/s]**********max output length : 26**********
total 0 cannot be tokenized
[0, 480, 6274, 17, 1]
[0, 480, 23766, 10536, 32, 27, 5605, 23, 1]
[0, 480, 23766, 16942, 32, 1138, 7, 11259, 1]
[0, 480, 22061, 77, 9, 5648, 11292, 9, 1]
[0, 480, 22061, 77, 9, 5648, 17, 17, 76, 1]
[0, 480, 22061, 24723, 1]
Using 16bit native Automatic Mixed Precision (AMP)
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
setup data for each process...
finished data processing... costing 0.3418557643890381s...
/home/joeysmith/miniconda3/envs/lambdakg/lib/python3.8/site-packages/pytorch_lightning/callbacks/model_checkpoint.py:611: UserWarning: Checkpoint directory output/metaQA exists and is not empty.
  rank_zero_warn(f"Checkpoint directory {dirpath} exists and is not empty.")
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
/home/joeysmith/miniconda3/envs/lambdakg/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py:2109: LightningDeprecationWarning: `Trainer.num_gpus` was deprecated in v1.6 and will be removed in v1.8. Please use `Trainer.num_devices` instead.
  rank_zero_deprecation(
/home/joeysmith/miniconda3/envs/lambdakg/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py:2077: LightningDeprecationWarning: `Trainer.num_processes` is deprecated in v1.6 and will be removed in v1.8. Please use `Trainer.num_devices` instead.
  rank_zero_deprecation(
/home/joeysmith/miniconda3/envs/lambdakg/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py:2093: LightningDeprecationWarning: `Trainer.tpu_cores` is deprecated in v1.6 and will be removed in v1.8. Please use `Trainer.num_devices` instead.
  rank_zero_deprecation(

  | Name  | Type        | Params
--------------------------------------
0 | model | T5KBQAModel | 222 M 
--------------------------------------
222 M     Trainable params
0         Non-trainable params
222 M     Total params
445.767   Total estimated model params size (MB)
/home/joeysmith/miniconda3/envs/lambdakg/lib/python3.8/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:240: PossibleUserWarning: The dataloader, train_dataloader, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 24 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
/home/joeysmith/miniconda3/envs/lambdakg/lib/python3.8/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:240: PossibleUserWarning: The dataloader, val_dataloader 0, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 24 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
Epoch 0:   0%|                                                       | 0/1501 [00:00<?, ?it/s]/home/joeysmith/miniconda3/envs/lambdakg/lib/python3.8/site-packages/torch/optim/lr_scheduler.py:129: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  warnings.warn("Detected call of `lr_scheduler.step()` before `optimizer.step()`. "
Epoch 1:  91%|███████████████████▉  | 1501/1657 [01:53<00:11, 13.21it/s, loss=0.651, v_num=21/home/joeysmith/miniconda3/envs/lambdakg/lib/python3.8/site-packages/transformers/generation_utils.py:1818: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  next_indices = next_tokens // vocab_size
Epoch 9: 100%|█| 1657/1657 [03:41<00:00,  7.49it/s, loss=0.506, v_num=21, hits1=0.000, hits3=0/home/joeysmith/miniconda3/envs/lambdakg/lib/python3.8/site-packages/pytorch_lightning/trainer/connectors/accelerator_connector.py:455: UserWarning: The flag `devices=1` will be ignored, instead the device specific number 1 will be used
  rank_zero_warn(
Using 16bit native Automatic Mixed Precision (AMP)
Trainer already configured with model summary callbacks: [<class 'pytorch_lightning.callbacks.model_summary.ModelSummary'>]. Skipping setting a default `ModelSummary` callback.
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
setup data for each process...
finished data processing... costing 0.023950815200805664s...
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
/home/joeysmith/miniconda3/envs/lambdakg/lib/python3.8/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:240: PossibleUserWarning: The dataloader, test_dataloader 0, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 24 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
Testing DataLoader 0: 100%|█████████████████████████████████| 156/156 [01:47<00:00,  1.45it/s]──────────────────────────────────────────────────────────────────────────────────────────────       Test metric             DataLoader 0
──────────────────────────────────────────────────────────────────────────────────────────────          hits1                     0.0
         hits10                     0.0
          hits3                     0.0
──────────────────────────────────────────────────────────────────────────────────────────────[{'hits1': 0.0, 'hits3': 0.0, 'hits10': 0.0}]

**Comment by zxlzr** [comment]
请问是否有改过超参数脚本

**Comment by JoeySmith1103** [comment]
除了前面提到的因为跑不动num_worker改为0，剩下的其他部分都没有改动。

**Comment by zxlzr** [comment]
您好抱歉回复晚了，我们尽快check下看能否复现您的问题

**Comment by CheaSim** [comment]
不好意思，由于Google T5模型权重存在一些pytorch兼容问题，所以只能使用32 全精度进行训练， 您只需要注释掉`--precision 16`即可。

**Comment by CheaSim** [comment]
您好，不好意思回复晚了。可以使用以下脚本，由于Google T5模型权重转换有问题，所以需要全精度训练（删除`precision 0`），同时将prefix tree设置为0，运行脚本如下`prefix_tree_decode 0`, 
```shell
CUDA_VISIBLE_DEVICES=0 python main.py  --max_epochs=10  --num_workers=0 \
   --model_name_or_path  t5-base \
   --num_sanity_val_steps 0 \
   --model_class T5KBQAModel \
   --prefix_tree_decode 0 \
   --lit_model_class KGT5LitModel \
   --label_smoothing 0.1 \
   --data_class MetaQADataModule \
   --devices 1 \
   --batch_size 64 \
   --check_val_every_n_epoch 2 \
   --dataset metaQA \
   --k_hop 1 \
   --eval_batch_size 128 \
   --max_seq_length 64 \
   --max_entity_length 128 \
   --lr 5e-5
```

**Comment by zxlzr** [comment]
请问您的问题是否已解决？

**Comment by JoeySmith1103** [comment]
已经解决，虽然目前看下来hist1执行到第7个epoch还只有0.07，但可能是batch size设定太小的问题，我再尝试看看，谢谢!

**Comment by zxlzr** [comment]
请问您还有其他问题吗?
